File: ios/Matcha/Matcha/MatchaTextView.m
#import "MatchaTextView.h"
#import "MatchaViewController.h"

@implementation MatchaTextView

+ (void)load {
    [MatchaViewController registerView:@"gomatcha.io/matcha/view/textview" block:^(MatchaViewNode *node){
        return [[MatchaTextView alloc] initWithViewNode:node];
    }];
}

- (id)initWithViewNode:(MatchaViewNode *)viewNode {
    if ((self = [super initWithFrame:CGRectZero])) {
        self.viewNode = viewNode;
    }
    return self;
}

- (void)setNativeState:(NSData *)nativeState {
    MatchaPBStyledText *text = [MatchaPBStyledText parseFromData:nativeState error:nil];
    NSAttributedString *attrString = [[NSAttributedString alloc] initWithProtobuf:text];
    self.attributedText = attrString;
    self.numberOfLines = 0;
}

@end

File: android/MatchaLib/matcha/src/main/java/io/gomatcha/matcha/MatchaTextView.java
package io.gomatcha.matcha;

import android.content.Context;
import android.text.SpannableString;
import android.widget.TextView;

import com.google.protobuf.InvalidProtocolBufferException;

import io.gomatcha.matcha.proto.text.PbText;

class MatchaTextView extends MatchaChildView {
    TextView view;

    static {
        MatchaView.registerView("gomatcha.io/matcha/view/textview", new MatchaView.ViewFactory() {
            @Override
            public MatchaChildView createView(Context context, MatchaViewNode node) {
                return new MatchaTextView(context, node);
            }
        });
    }

    public MatchaTextView(Context context, MatchaViewNode node) {
        super(context, node);

        view = new TextView(context);
        addView(view);
    }

    @Override
    public void setNativeState(byte[] nativeState) {
        super.setNativeState(nativeState);
        try {
            PbText.StyledText proto  = PbText.StyledText.parseFrom(nativeState);
            SpannableString str = Protobuf.newAttributedString(proto);
            view.setText(str);
        } catch (InvalidProtocolBufferException e) {
        }
    }
}
